File: solutions/go/03/abc137_d/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(in, &n)
	jobs := make([][2]int, n)
	maxd := 0
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &jobs[i][0], &jobs[i][1])
		if jobs[i][0] > maxd {
			maxd = jobs[i][0]
		}
	}
	sort.Slice(jobs, func(i, j int) bool {
		return jobs[i][0] < jobs[j][0]
	})
	dp := make([]int, maxd+2)
	for _, job := range jobs {
		d, p := job[0], job[1]
		for i := maxd; i >= d; i-- {
			if dp[d-1]+p > dp[i] {
				dp[i] = dp[d-1] + p
			}
		}
	}
	ans := 0
	for _, v := range dp {
		if v > ans {
			ans = v
		}
	}
	fmt.Println(ans)
}

File: solutions/python/03/abc137_d/main.py
import sys


def main() -> None:
    input = sys.stdin.readline
    n = int(input())
    jobs = []
    for _ in range(n):
        d, p = map(int, input().split())
        jobs.append((d, p))
    jobs.sort()
    maxd = max(d for d, _ in jobs)
    dp = [0] * (maxd + 2)
    for d, p in jobs:
        for i in range(maxd, d - 1, -1):
            dp[i] = max(dp[i], dp[d - 1] + p)
    print(max(dp))


if __name__ == "__main__":
    main()
